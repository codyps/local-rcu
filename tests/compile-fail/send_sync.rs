// Check that non-sync types can't be sent between threads via rcu
// edition:2021

use std::cell::RefCell;
use std::thread;

// NOTE: we include this so the error points to local source which gives reliable trybuild output
fn spawn<F>(f: F)
where
    F: FnOnce() + Send + 'static,
{
    thread::spawn(f).join().unwrap();
}

fn main() {
    let (w, mut r1) = local_rcu::slot(RefCell::new(0));

    let t1 = spawn(move || {
        *r1.read().borrow_mut() = 1;
    });
    //~^ ERROR: `RefCell<i32>` cannot be shared between threads safely

    assert_eq!(*w.read().borrow(), 0);
}
