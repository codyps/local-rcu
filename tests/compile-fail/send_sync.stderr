error[E0277]: `RefCell<i32>` cannot be shared between threads safely
  --> tests/compile-fail/send_sync.rs:18:20
   |
18 |       let t1 = spawn(move || {
   |  ______________-----_^
   | |              |
   | |              required by a bound introduced by this call
19 | |         *r1.read().borrow_mut() = 1;
20 | |     });
   | |_____^ `RefCell<i32>` cannot be shared between threads safely
   |
   = help: the trait `Sync` is not implemented for `RefCell<i32>`, which is required by `{closure@$DIR/tests/compile-fail/send_sync.rs:18:20: 18:27}: Send`
   = note: if you want to do aliasing and mutation between multiple threads, use `std::sync::RwLock` instead
   = note: required for `Reader<RefCell<i32>>` to implement `Send`
note: required because it's used within this closure
  --> tests/compile-fail/send_sync.rs:18:20
   |
18 |     let t1 = spawn(move || {
   |                    ^^^^^^^
note: required by a bound in `spawn`
  --> tests/compile-fail/send_sync.rs:10:19
   |
8  | fn spawn<F>(f: F)
   |    ----- required by a bound in this function
9  | where
10 |     F: FnOnce() + Send + 'static,
   |                   ^^^^ required by this bound in `spawn`
